enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,456 @@
+   to ensure no tests' dependency affects the results.
+3. The next part of `Run` method is responsible for:
+   - filtering tests based on API groups;
+   - setting up various monitors, which gather data about cluster under test;
+   - splitting the tests into several buckets such as early, late, and several others,
+     which help with parallelization of the execution.
+4. The actual logic responsible for executing tests is located [here](https://github.com/openshift/origin/blob/cc42164781a728b804d0bb07f09cf878ec8f7807/pkg/test/ginkgo/cmd_runsuite.go#L367-L402),
+   it places all items in the [queue](https://github.com/openshift/origin/blob/master/pkg/test/ginkgo/queue.go),
+   and [executes tests in a new process](https://github.com/openshift/origin/blob/cc42164781a728b804d0bb07f09cf878ec8f7807/pkg/test/ginkgo/test_runner.go#L296)
+   invoking `<test-binary> run-test <test-name>`.
+5. The reminder of the `Run` method is responsible for:
+   - analyzing the monitor data;
+   - summarizing the tests results.
+
+### Future work
+
+The current state of art, as mentioned above, provided mechanism for both
+extracting the test binaries, and using those binaries for running tests.
+The followup work requires:
+
+1. Extracting common test library for re-use. The library should be based on the
+   contents of [k8s-tests directory](https://github.com/openshift/kubernetes/tree/8bd5514be14de7fb3105962c8fc206ef5e921a0b/openshift-hack/cmd/k8s-tests).
+2. Expanding [externalTestsForSuite](https://github.com/openshift/origin/blob/cc42164781a728b804d0bb07f09cf878ec8f7807/pkg/test/ginkgo/external.go#L29)
+   to allow extracting test binaries from other images, currently it uses hardcoded
+   `hyperkube` image from the release payload, but should rather rely on [an annotation](#implementation-detailsnotesconstraints),
+   telling which images contain the external test binary, and its location inside
+   the image.
+3. Preparing a template/demo for one of operators building on top of steps 1 and 2
+   above.
+
+Additionally, to enable flows where release payload is not available, one should be
+able to point to a local directory containing external binaries that should be
+used instead. For example, through `OPENSHIFT_EXTERNAL_TESTS_DIRECTORY` environment
+variable.
+
+### Even more future work
+
+As mentioned in [risks section](#risks-and-mitigations), upgrade tests are relying
+on the vendored code only, it would be reasonable to address this issue in the long,
+to run the upgrade tests in a similar fashion as the other tests.
+
+Re-evaluate [goal #4](#goals) which explicitly requires providing a fallback for
+kubernetes vendored tests. Currently, at least Microshift requires us to have
+the kubernetes tests vendored since they don't provide release image from which
+we can extract the kubernetes test binaries.
+
+Re-evaluate [alternative approach](#build-time-assembly---aggregating-binaries)
```

Comment:
It's probably worth starting this conversation with testplatform in 4.15 to understand what capabilities our CI build system provides and how difficult this actually is long term.  I'd really like to see us not have vendored tests.  That will discourage separation beyond a few well known cases.